Merge #40878: Add regexp tree dictionary

hanfei1991 opened the pull request:
### Changelog category (leave one):
- New Feature


### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
to merge #40878 , supporting regexp dictionary.

The YAML source has a new format. For every yaml node, the format will be like:
- key_name (defined by dictioanary structure) : regular expression
- attribute_name (defined by dictionary structure) : value (might be a string containing back-reference)
- ... (multiple attribute names)
- name which is not key or attributes: another yaml node (for recursive match)

This format is to keep compatible with: https://github.com/ua-parser/uap-core and  https://github.com/matomo-org/device-detector/tree/master/regexes

The match order is : 
1. son is higher than parents, 
2. smaller id is highter than larger id if they are in the same layer. 


> Information about CI checks: https://clickhouse.com/docs/en/development/continuous-integration/

nooblose committed:
Add YAML regular expressions tree dictionary source

nooblose committed:
Add RegExpTree dictionary

vdimir committed:
dummy commit

hanfei1991 committed:
Merge branch 'master' into regexp-tree-dictionary

hanfei1991 committed:
fix compile

hanfei1991 committed:
fix macro flag

hanfei1991 committed:
fix check-style

hanfei1991 committed:
fix style check

hanfei1991 committed:
remove integration tests to check CI

hanfei1991 committed:
reimplement

hanfei1991 committed:
Merge branch 'master' into regexp-tree-dictionary

hanfei1991 committed:
support regex tree dictionary

hanfei1991 committed:
fix build

hanfei1991 committed:
fix build

hanfei1991 committed:
use vector scan in test

hanfei1991 committed:
add test for yaml source

hanfei1991 commented:
Next goal is:
1. Support popular ua-parser lib his format is to keep compatible with: https://github.com/ua-parser/uap-core and https://github.com/matomo-org/device-detector/tree/master/regexes , the problem with them right now is:
    - the regular expressions in the uap-core include finite boundries like "{0,200}" which cause huge pressure on vectorscan. we should refine these re before we test this lib
    - the regular expressions in device-detector contain `zero-length assertions` which is not supported by vectorscan. I will see if it's possible to avoid this syntax.
2. Support more types parsing in yaml file and handle exceptions for parse error.

The other parts are ready to review. PTAL @vdimir

vdimir commented on src/Dictionaries/YAMLRegExpTreeDictionarySource.cpp:
```diff
@@ -0,0 +1,350 @@
+
+QueryPipeline YAMLRegExpTreeDictionarySource::loadAll()
+{
+    LOG_INFO(logger, "Loading all from {}", filepath);
```
```suggestion
    LOG_INFO(logger, "Loading regexp tree from yaml '{}'", filepath);
```

vdimir commented on src/Dictionaries/RegExpTreeDictionary.cpp:
```diff
@@ -0,0 +1,545 @@
+        regexp_options.set_log_errors(false);
+        RegexTreeNodePtr node = std::make_unique<RegexTreeNode>(id, parent_id, regex, regexp_options);
+
+        int num_captures = std::min(node->searcher.NumberOfCapturingGroups() + 1, 10);
```
Why 10 is the minimum?

hanfei1991 replied:
just copy from https://github.com/ClickHouse/ClickHouse/blob/150a699dda1033a35f267689ece8f81f1cc672e6/src/Functions/ReplaceRegexpImpl.h#L49
and in uap-core the back-ref is also $1-$9, so I think this restraint is reasonable.

vdimir commented on src/Dictionaries/RegExpTreeDictionary.cpp:
```diff
@@ -0,0 +1,545 @@
+    {
+        Block sample_block;
+        /// id, parent_id, regex, keys, values
+        sample_block.insert(ColumnWithTypeAndName(DataTypePtr(new DataTypeUInt64()), kId));
+        sample_block.insert(ColumnWithTypeAndName(DataTypePtr(new DataTypeUInt64()), kParentId));
+        sample_block.insert(ColumnWithTypeAndName(DataTypePtr(new DataTypeString()), kRegExp));
+        sample_block.insert(ColumnWithTypeAndName(DataTypePtr(new DataTypeArray(DataTypePtr(new DataTypeString()))), kKeys));
+        sample_block.insert(ColumnWithTypeAndName(DataTypePtr(new DataTypeArray(DataTypePtr(new DataTypeString()))), kValues));
```
Consider using `std::make_shared<DataTypeArray>(std::make_shared<DataTypeString>())`

vdimir commented on src/Dictionaries/RegExpTreeDictionary.cpp:
```diff
@@ -0,0 +1,545 @@
+    for (const auto & item : pieces)
+    {
+        if (item.ref_num >= 0)
+            result += matches[item.ref_num].ToString();
```
We already checked `ref_num` in `createStringPieces`, but let's also add check here to be sure that `item.ref_num < 10` and we don't overflow `matches` array

vdimir commented on src/Dictionaries/RegExpTreeDictionary.cpp:
```diff
@@ -0,0 +1,545 @@
+    const StorageID & id_, const DictionaryStructure & structure_, DictionarySourcePtr source_ptr_, Configuration configuration_)
+    : IDictionary(id_), structure(structure_), source_ptr(source_ptr_), configuration(configuration_), logger(&Poco::Logger::get("RegExpTreeDictionary"))
+{
+    if (auto * ch_source = typeid_cast<ClickHouseDictionarySource *>(source_ptr.get()))
```
How sample_block will work if source is other than `Clickhouse` ? Is it safe to change `ClickHouseDictionarySource::sample_block` after `ClickHouseDictionarySource` was created?

hanfei1991 replied:
if source is not `ClickHouse`, the schema will mismatch and raise a Exception.

This is only for test. We assume that the user only uses YAML source by default.

vdimir commented on tests/queries/0_stateless/02504_regexp_dictionary_table_source.sql:
```diff
@@ -0,0 +1,81 @@
+    comment String default 'nothing'
+)
+PRIMARY KEY(regexp)
+SOURCE(CLICKHOUSE(QUERY concat('select * from ', currentDatabase() , '.regexp_dictionary_source_table')))
```
Why it's not just `SOURCE(CLICKHOUSE(TABLE 'regexp_dictionary_source_table'))`?

hanfei1991 replied:
because the schema will mismatch if we refer a TABLE. Though if we refer a QUERY, the schema will be decided by sample_block. That's trikcy but I think it's ok for test.

vdimir commented on src/Dictionaries/RegExpTreeDictionary.cpp:
```diff
@@ -0,0 +1,545 @@
+    if (auto * ch_source = typeid_cast<ClickHouseDictionarySource *>(source_ptr.get()))
+    {
+        Block sample_block;
+        /// id, parent_id, regex, keys, values
```
How exactly existing sample_block differ? In tests we have the same
```sql
CREATE TABLE regexp_dictionary_source_table
(
    id UInt64,
    parent_id UInt64,
    regexp String,
    keys   Array(String),
    values Array(String),
```

hanfei1991 replied:
Yeah but the schema of Dictionary is (regexp, attribute_1, attribute_2 ...) . ClickHouse assumes dictionary and ClickHouse source have same schema

vdimir commented on src/Dictionaries/YAMLRegExpTreeDictionarySource.cpp:
```diff
@@ -0,0 +1,350 @@
+*   - regexp: "MSIE (\\d+)"
+*     attr1: "Vecna"
+*     attr2: 22.8
+*     match: # nested match for subpattern
```
in tests it's named `versions`, so what is the correct way to define subpath in this tree?

hanfei1991 replied:
indeed, the subpath attrubute name can be arbitrary.

vdimir reviewed (commented):
Sorry for the delay. 
The code almost looks good, just some comments. Also, It's worth adding some docs (in this or followup PR).

hanfei1991 committed:
Update tests/queries/0_stateless/02504_regexp_dictionary_yaml_source.sh

Co-authored-by: Vladimir C <vdimir@clickhouse.com>

hanfei1991 committed:
Update tests/queries/0_stateless/02504_regexp_dictionary_yaml_source.sh

Co-authored-by: Vladimir C <vdimir@clickhouse.com>

hanfei1991 committed:
address comments

hanfei1991 committed:
add check sanitizer

hanfei1991 committed:
try to fix flaky tests

hanfei1991 committed:
Merge branch 'master' into regexp-tree-dictionary

hanfei1991 commented:
Future jobs:
- [ ] write docs
- [ ] catch exception for parse literal error, for example, type int64 meets 'abc'
- [ ] restrict source for regex dict (only allow yaml source, add a setting for clickhouse table source)
- [ ] restrict data types for regex dict (array / nested columns might be not appropriate)
jobs after this pr:
- [ ] support uap-core for ua parsing and add it to test
- [ ] support low-cardinality columns
- [ ] support cache

hanfei1991 committed:
fix style check

hanfei1991 committed:
some improvement

hanfei1991 committed:
fix build

hanfei1991 committed:
refine test

hanfei1991 committed:
write docs and optimize regex compile

hanfei1991 committed:
fix check style

hanfei1991 committed:
fix build

hanfei1991 committed:
fix build

hanfei1991 committed:
fix build

hanfei1991 commented:
@vdimir PTAL. I tested more data types and added restrictions for source types. Now it's safer than before. I also added some simple docs with simple examples. Not thouroughly but enough for a good begin :)

vdimir commented on docs/en/sql-reference/dictionaries/external-dictionaries/regexp-tree.md:
```diff
@@ -0,0 +1,76 @@
+---
+import CloudDetails from '@site/docs/en/sql-reference/dictionaries/external-dictionaries/_snippet_dictionary_in_cloud.md';
+
+Regexp Tree dictionary stores multiple trees of regular expressions with attributions. Users can trieve strings in the dictionary. If a string matches the root of the regexp tree, we will collect the corresponding attributions of the matched root and continue to walk the children. If any of the children matches the string, we will collect attributions and rewrite the old ones if conflicts occur,  ...
```
Some typos

```suggestion
Regexp Tree dictionary stores multiple trees of regular expressions with attributions. Users can retrieve strings in the dictionary. If a string matches the root of the regexp tree, we will collect the corresponding attributes of the matched root and continue to walk the children. If any of the children matches the string, we will collect attributes and rewrite the old ones if conflicts occur, then continue the traverse until we reach leaf nodes.
```

vdimir reviewed (approved):
LGTM

hanfei1991 committed:
Update docs/en/sql-reference/dictionaries/external-dictionaries/regexp-tree.md

Co-authored-by: Vladimir C <vdimir@clickhouse.com>

hanfei1991 committed:
Merge branch 'master' into regexp-tree-dictionary

hanfei1991 committed:
Merge branch 'master' into regexp-tree-dictionary

tavplubix commented on src/Dictionaries/RegExpTreeDictionary.cpp:
```diff
@@ -0,0 +1,554 @@
+    for (const auto & [id, value]: regex_nodes)
+        if (value->parent_id == 0) // this is root node.
+            initTopologyOrder(id, visited, topology_id);
+    if (topology_order.size() != regex_nodes.size())
+        throw Exception(ErrorCodes::INCORRECT_DICTIONARY_DEFINITION, "Invalid Regex tree");
```
This exception message is not really informative for a user, could you please update it (or replace with `LOGICAL ERROR` if it's not for a user)? (what is invalid? how to fix it?)

hanfei1991 replied:
actually this belongs to LOGICAL_ERROR. fixing in https://github.com/ClickHouse/ClickHouse/pull/48632
Thank you!

tavplubix commented on src/Dictionaries/RegExpTreeDictionary.cpp:
```diff
@@ -0,0 +1,554 @@
+{
+    visited.insert(node_idx);
+    for (UInt64 child_idx : regex_nodes[node_idx]->children)
+        if (visited.contains(child_idx))
+            throw Exception(ErrorCodes::INCORRECT_DICTIONARY_DEFINITION, "Invalid Regex tree");
```
Same
